adaptivemm/src/adaptivemmd.c

```diff
@@ -519,6 +535,93 @@ void rescale_maxwsf()
 		log_warn("Failed to compute reasonable WSF, %ld, total pages %ld, reclaimable pages %ld", new_wsf, total_managed, reclaimable_pages);
 }
 
+int parse_fs_files(char * path, unsigned long *data)
+{
+	FILE *fp = NULL;
+	char line[80];
+
+	fp = fopen(path, "r");
+	if (!fp)
+		return -1;
+
+	if (fgets(line, sizeof(line), fp) != NULL) {
+		sscanf(line, "%lu\n", data);
+	} else
+		return -1;
+
+	return 0;
+}
+
+int set_vfs_cache_pressure(unsigned long new_cache_pressure)
+{
+	int fd;
+	char scaled_cache_pressure[20];
+	printf("set_vfs_cache_pressure = %ld", new_cache_pressure);
+
+	sprintf(scaled_cache_pressure, "%ld\n", new_cache_pressure);
+	if ((fd = open(VFS_CACHE_PRESSURE, O_WRONLY)) == -1) {
+		log_err("Failed to open "VFS_CACHE_PRESSURE" (%s)", strerror(errno));
+		return -1;
+	}
+
+	if (write(fd, scaled_cache_pressure, strlen(scaled_cache_pressure)) < 0) {
+		log_err("Failed to write to "VFS_CACHE_PRESSURE" (%s)", strerror(errno));
+		return -1;
+	}
+
+	log_info(1, "New vfs_cache_pressure = %ld", new_cache_pressure);
+	return 0;
+}
+
+/*
+ * rescale vm.vfs_cache_pressure vaue based on current memory trends
+ */
+void rescale_vfs_cache_pressure()
+{
+	struct timespec spec;
+	unsigned long curr_inodes;
+	unsigned long curr_dentries;
+	unsigned long curr_vfs_cache_pressure;
+	long long dentry_m, inode_m;
+	long long dentry_c, inode_c;
+
+	if (parse_fs_files(VFS_CACHE_PRESSURE, &curr_vfs_cache_pressure) ||
+		parse_fs_files(INODEINFO, &curr_inodes) ||
+		parse_fs_files(DENTRYINFO, &curr_dentries))
+			return;
+
+	clock_gettime(CLOCK_MONOTONIC_RAW, &spec);
+
+	if(lsq_fit(&fs_lsq[0], curr_dentries,
+				(long long)get_msecs(&spec),&dentry_m, &dentry_c) ||
+		lsq_fit(&fs_lsq[1], curr_inodes,
+				(long long)get_msecs(&spec), &inode_m, &inode_c))
+			return;
+	/*
+	 * adjust vm.vfs_cache_pressure based on the
+	 * growth of cached inode and dentry objects
+	 */
+	if (prefer_object_caching) {
+		if ((dentry_m > 0 || inode_m > 0) &&
+			curr_vfs_cache_pressure > MIN_VFS_CACHE_PRESSURE) {
+			set_vfs_cache_pressure(curr_vfs_cache_pressure * 0.9);
+		}
+		else if ((dentry_m < 0 || inode_m < 0) &&
+			curr_vfs_cache_pressure < MAX_VFS_CACHE_PRESSURE/1.1)
+			set_vfs_cache_pressure(curr_vfs_cache_pressure * 1.1);
```

Comment:
Do you need/want to do a `min()` here to ensure that we don't set a value above the `MAX_VFS_CACHE_PRESSURE`?